### PR TITLE
Update pydantic in apps/hermes

### PR DESCRIPTION
**Pull request by islean**
## Description

This updates the cg/apps/hermes directory and corresponding tests to use pydantic v2. For documentation on annotated validators, see https://docs.pydantic.dev/latest/usage/validators/#annotated-validators.

### Changed

- Hermes uses pydantic 2 for its models.
- Tests for hermes uses pydantic 2 for its ValidationErrors.

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b update-pydantic-hermes -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/apps/hermes/models.py [hunk 1]
```diff
@@ -16,23 +17,24 @@ class CGTag(BaseModel):
     mandatory: Optional[bool] = False
 
 
+def check_mandatory_exists(file: CGTag):
```
Comment (henrikstranneheim) [inline cg/apps/hermes/models.py:20]: Add doc str

File: cg/apps/hermes/models.py [hunk 2]
```diff
@@ -16,23 +17,24 @@ class CGTag(BaseModel):
...
+    if file.mandatory:
+        assert Path(file.path).exists(), f"Mandatory file cannot be found at {file.path}"
```
Comment (henrikstranneheim) [inline cg/apps/hermes/models.py:22]: I would like to avoid using "assert" in production code.

File: cg/apps/hermes/models.py [hunk 3]
```diff
@@ -16,23 +17,24 @@ class CGTag(BaseModel):
...
+        return file
+    if not Path(file.path).exists():
+        LOG.info("Optional file %s not found, removing from bundle", file.path)
```
Comment (henrikstranneheim) [inline cg/apps/hermes/models.py:25]: Use f-str

File: cg/apps/hermes/models.py (hunk 1 above)
Comment (henrikstranneheim) [inline cg/apps/hermes/models.py:20]:
```suggestion
def check_mandatory_exists(file: CGTag) -> Optional[CgTag]:
```

File: cg/apps/hermes/models.py [hunk 4]
```diff
@@ -16,23 +17,24 @@ class CGTag(BaseModel):
...
-            LOG.info("Optional file %s not found, removing from bundle", file.path)
-            return
-        return file
+    files: List[Annotated[CGTag, AfterValidator(check_mandatory_exists)]]
 
-    @validator("files")
+    @classmethod
+    @field_validator("files")
     def remove_invalid(cls, value):
         return [item for item in value if item]
```
Comment (henrikstranneheim) [inline cg/apps/hermes/models.py:40]: Not strictly part of the Pydantic migration, but could you take the opportunity to improve the function name, signature and use more mnemonic variables

**Review by henrikstranneheim — APPROVED** [review]
👍

File: cg/apps/hermes/models.py (hunk 2 above)
Comment (islean) [reply cg/apps/hermes/models.py:22]: Seems to be the way validators are written - should we avoid them then? See e.g. https://docs.pydantic.dev/latest/usage/validators/

File: cg/apps/hermes/models.py (hunk 4 above)
Comment (islean) [reply cg/apps/hermes/models.py:40]: Sure thing!

Comment (islean) [reply cg/apps/hermes/models.py]: I have tried reworking this to a more understandable validation process now. Can you see if the current validation matches the old one? @henrikstranneheim

File: cg/apps/hermes/models.py (hunk 2 above)
Comment (henrikstranneheim) [reply cg/apps/hermes/models.py]: From Pydantic docs: Warning: If you make use of assert statements, keep in mind that running Python with the -O optimization flag disables assert statements, and validators will stop working.

I do not think it is worth the gain when it so easily replaced.

Comment (islean) [reply cg/apps/hermes/models.py]: Exchanged for a "raise error"

**Review by henrikstranneheim — APPROVED** [review]
👍

**Comment by islean** [comment]
Test performed by IO wherein the branch was deployed to stage and `cg workflow mip-dna store <case_internal_id>` was executed successfully (which instantiates the base models).
 `Analysis successfully stored in Housekeeper: <case_internal_id> : 2023-08-23 12:16:18`

**Pull request merged**